### ISSUE TITLE
Tokenize incorrect when getting pretrianed feature

### ISSUE THREAD

**Issue by Chris-Tang6** [seyonechithrananda/bert-loves-chemistry#60]
Hi ChemBERTa team🤗, I got a problem when I tokenize smiles seq, And I found that in you example smiles has the `Cl` atom. So I want to konw if you meet the same question.

The Iuput seq is ```COC1=C(C=C2C(=C1)CCN=C2C3=CC(=C(C=C3)Cl)Cl)Cl```
However the output of tokenizer incorrectly labeled the `Cl` as `C`, I don't konw how the fix it. And I also found that in the ChenBERTa token table, it indeed has the 'Cl' token.
![image](https://github.com/huggingface/transformers/assets/53926670/9877917a-a426-4901-8686-703dffb34853)

The output of the tokenize() result as follow:
```['C', 'O', 'C', '1', '=', 'C', '(', 'C', '=', 'C', '2', 'C', '(', '=', 'C', '1', ')', 'C', 'C', 'N', '=', 'C', '2', 'C', '3', '=', 'C', 'C', '(', '=', 'C', '(', 'C', '=', 'C', '3', ')', 'C', ')', 'C', ')', 'C']```

_Originally posted by @Chris-Tang6 in https://github.com/seyonechithrananda/bert-loves-chemistry/issues/58#issuecomment-1784738131_

**Comment by MrsW6** [comment]
yes，same question

**Comment by Lawwwwo** [comment]
same question for me, not only 'Cl' to 'C', like '[Zn]' to 'n' etc...

**Comment by kc-chong** [comment]
Hi I find the toknenizers to be behaving quite differently across different models, and also observed its incorrect parsing of CL. I tested all the available models on HuggingFace using my internal dataset. Here's the results, the tokenizers of smaller sets seem to behave much better. And embrassingly all the tokenizers under DeepChem directory had issue with tokenization. Hope this helps. 

| Model Name                                      | CL Encoding | [Zn] Encoding       | Recovery Rate |
|-------------------------------------------------|-------------|---------------------|---------------|
| seyonec/PubChem10M_SMILES_BPE_450k              | [Cl]        | [[, Zn, ]]         | 1.0           |
| seyonec/ChemBERTa_zinc250k_v2_40k               | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/ChemBERTa-zinc-base-v1                  | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/SMILES_tokenized_PubChem_shard00_160k   | [C]         | [n]                | 0.087146      |
| seyonec/PubChem10M_SMILES_BPE_60k               | [Cl]        | [[, Zn, ]]         | 1.0           |
| seyonec/SMILES_tokenized_PubChem_shard00_50k    | Failed      | Failed             | Failed        |
| seyonec/SMILES_BPE_PubChem_100k_shard00         | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/PubChem10M_SMILES_BPE_396_250           | [Cl]        | [[, Zn, ]]         | 1.0           |
| seyonec/ChemBERTa-zinc250k-v1                   | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/SMILES_BPE_PubChem_250k                 | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/PubChem10M_SMILES_BPE_240k              | [Cl]        | [[, Zn, ]]         | 1.0           |
| seyonec/ChemBERTA_PubChem1M_shard00_155k        | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/ChemBERTA_PubChem1M_shard00_140k        | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/ChemBERTA_PubChem1M_shard00_125k        | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/PubChem10M_SMILES_BPE_390k              | [Cl]        | [[, Zn, ]]         | 1.0           |
| seyonec/BPE_SELFIES_PubChem_shard00_160k        | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/ChemBERTa_masked_30_PubChem_shard00_1M_150k_steps | [Cl] | [[, Z, n, ]] | 1.0 |
| seyonec/SmilesTokenizer_ChemBERTa_zinc250k_40k  | Failed      | Failed             | Failed        |
| seyonec/SMILES_tokenized_PubChem_shard00_150k   | Failed      | Failed             | Failed        |
| seyonec/BPE_SELFIES_PubChem_shard00_120k        | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/ChemBERTA_PubChem1M_shard00_115k        | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/BPE_SELFIES_PubChem_shard00_166_5k      | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/BPE_SELFIES_PubChem_shard00_50k         | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/BPE_SELFIES_PubChem_shard00_70k         | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/ChemBERTA_PubChem1M_shard00_75k         | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/PubChem10M_SMILES_BPE_120k              | [Cl]        | [[, Zn, ]]         | 1.0           |
| seyonec/SMILES_tokenized_PubChem_shard00_100k   | Failed      | Failed             | Failed        |
| seyonec/SMILES_tokenized_PubChem_shard00_40k    | Failed      | Failed             | Failed        |
| seyonec/BPE_SELFIES_PubChem_shard00_150k        | [Cl]        | [[, Z, n, ]]       | 1.0           |
| seyonec/PubChem10M_SMILES_BPE_180k              | [Cl]        | [[, Zn, ]]         | 1.0           |
| seyonec/PubChem10M_SMILES_BPE_50k               | [Cl]        | [[, Zn, ]]         | 1.0           |
| DeepChem/ChemBERTa-10M-MTR                      | [C]         | [n]                | 0.087146      |
| DeepChem/ChemBERTa-5M-MTR                       | [C]         | [n]                | 0.087146      |
| DeepChem/ChemBERTa-77M-MTR                      | [C]         | [n]                | 0.087146      |
| DeepChem/ChemBERTa-10M-MLM                      | [C]         | [n]                | 0.087146      |
| DeepChem/ChemBERTa-5M-MLM                       | [C]         | [n]                | 0.087146      |
| DeepChem/ChemBERTa-77M-MLM                      | [C]         | [n]                | 0.087146      |
| DeepChem/SmilesTokenizer_PubChem_1M             | [C]         | [n]                | 0.087146      |

**Comment by DreRnc** [comment]
I think the tokenizer is broken because the vocabulary expects inputs that the pre-tokenizer (which breaks up the string) does not yield. I'll try to explain this through a simple example (with code). 

Let's say I have the SMILES "[O-][N+]1=CC=CC=C1"

I would want the tokenizer to convert into IDs. When using the "DeepChem/ChemBERTa-77M-MLM" tokenizer, the intended behavior would be that the vocabulary transforms the tokens into the following IDs.
[O-] --> 36
[N+] --> 41
...

However, this conversion simply cannot happen because the pre-tokenizer is not giving as output "[O-]", "[N+]", ... and is breaking up the smiles differently, according to certain standard delimiters or patterns (e.g., special characters like -, =, or numbers). In particular, the pre-tokenizer would not have any chemical knowledge to split the SMILES in an intelligent way.

So the tokenization scheme seems to be broken. I'm wondering if the model is trained on this scheme, or if we are simply using incorrect settings.

To give an example, to explain what I mean:

```
from transformers import AutoTokenizer, AutoModelForMaskedLM

tokenizer = AutoTokenizer.from_pretrained("DeepChem/ChemBERTa-77M-MLM")
model = AutoModelForMaskedLM.from_pretrained("DeepChem/ChemBERTa-77M-MLM")

def visualize_smiles(smiles):

    pre_tokens = tokenizer.backend_tokenizer.pre_tokenizer.pre_tokenize_str(smiles)
    toks = tokenizer(smiles)
    tokens = tokenizer.convert_ids_to_tokens(toks["input_ids"])

    print("Pre_tokens: ", pre_tokens)
    print("Token_ids: ", toks["input_ids"])
    print("Tokens: ", tokens)
    
smiles = "[O-][N+]1=CC=CC=C1"

visualize_smiles(smiles)
```

```
Pre_tokens:  [('[', (0, 1)), ('O', (1, 2)), ('-][', (2, 5)), ('N', (5, 6)), ('+]', (6, 8)), ('1', (8, 9)), ('=', (9, 10)), ('CC', (10, 12)), ('=', (12, 13)), ('CC', (13, 15)), ('=', (15, 16)), ('C', (16, 17)), ('1', (17, 18))]
Token_ids:  [12, 19, 31, 23, 20, 22, 16, 16, 22, 16, 16, 22, 16, 20, 13]
Tokens:  ['[CLS]', 'O', '-', 'N', '1', '=', 'C', 'C', '=', 'C', 'C', '=', 'C', '1', '[SEP]']
```

The tokenizer could never have output the IDs of [O-],[N+],... because the pre-tokenizer broke the string in the following way:
[, O, -][, N, +], 1, =, CC, =, CC, C, 1

If this is how the actual tokenization is happening, it would not be making correct use of the vocabulary, which was very nicely constructed and has all the atoms and groups of characters that are chemically valid. At that point, it would be interesting to see if this is completely changing the model performance with respect to a tokenizer that is actually able to split the SMILES in the originally intended way (the one in the vocabulary).

I'll give the first 40 entries of the tokenizer.get_vocab() of DeepChem/ChemBERTa-77M-MLM to give an idea of how the (nice) intended tokenization scheme was supposed to be:

{'[PAD]': 0,
 '[unused1]': 1,
 '[unused2]': 2,
 '[unused3]': 3,
 '[unused4]': 4,
 '[unused5]': 5,
 '[unused6]': 6,
 '[unused7]': 7,
 '[unused8]': 8,
 '[unused9]': 9,
 '[unused10]': 10,
 '[UNK]': 11,
 '[CLS]': 12,
 '[SEP]': 13,
 '[MASK]': 14,
 'c': 15,
 'C': 16,
 '(': 17,
 ')': 18,
 'O': 19,
 '1': 20,
 '2': 21,
 '=': 22,
 'N': 23,
 '.': 24,
 'n': 25,
 '3': 26,
 'F': 27,
 'Cl': 28,
 '>>': 29,
 '~': 30,
 '-': 31,
 '4': 32,
 '[C@H]': 33,
 'S': 34,
 '[C@@H]': 35,
 '[O-]': 36,
 'Br': 37,
 '#': 38,
 '/': 39,
 '[nH]': 40,
...
}

Let us know if this is what is happening, or if I'm just using incorrect settings, thanks!

**Comment by DreRnc** [comment]
By the way, this can have even worse effects, such as leading to atoms being thrown out.

In the example of chlorophyll, for example, this leads to the Mg atom being totally ignored:

```
smiles = "C=Cc1c(C)c2n3c1C=C1C(C)=C(CC)C4=N1->[Mg]31<-N3=C(C5=c6c(c(C)c(n61)=C4)C(=O)C5C(=O)OC)C(CCC(=O)OC/C=C(\C)CCCC(C)CCCC(C)CCCC(C)C)C(C)C3=C2"

visualize_smiles(smiles)
```

output (truncated for "N1->[Mg]31<-N" for readability):

```
Pre_tokens:  [... , ('N', (0, 1)), ('1', (1, 2)), ('->[', (2, 5)), ('Mg', (5, 7)), (']', (7, 8)), ('31', (8, 10)), ('<-', (10, 12)), ('N', (12, 13)), ...]
Token_ids:  [... , 23, 20, 31, 26, 20, 31, 23, ...]
Tokens:  [... , 'N', '1', '-', '3', '1', '-', 'N', ...]
```

One can see how the pre-tokenizer spits out:
N, 1, ->[, Mg, ], 31, <-, N

But because the tokenizer doesn't know "Mg", but only "[Mg]", it will transform both "->[" and "Mg" to 31, which is "-". 

This got me wondering if, because of [,] making the pre-tokenizer split the strings, none of the groups "[...]" in the tokenizer's vocabulary are ever assigned.

I tested this by running the following code on ~90k mol objects, checking if there were any tokens (different from "[CLS]" and "[SEP]") that had a "[":

```
for mol in mols:
  smiles = Chem.MolToSmiles(mol)
  
  toks = tokenizer(smiles)
  tokens = tokenizer.convert_ids_to_tokens(toks["input_ids"])
  
  for token in tokens:
      assert ("[" not in token) or (token in ["[CLS]", "[SEP]"]), token
```

I never got an assertion error, meaning that the tokenizer never assigned any of the groups [...] (e.g., "[O-]", ["CH3+"], ...) in the vocabulary.

Again, let me know if there is something I am misunderstanding about the tokenizer!

**Comment by tspecht93** [comment]
Hi,

We encountered the same problem with the ChemBERTa models. Here is a hotfix that we used:

```
def initiliaze_ChemBERTA(model_name="DeepChem/ChemBERTa-77M-MTR", device=None):
    # Load the tokenizer from the pre-trained model
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    
    # Create the directory if it doesn't exist
    os.makedirs('ChemBERTa', exist_ok=True)
    
    # Save the tokenizer's vocabulary to the specified folder
    tokenizer.save_vocabulary('ChemBERTa/')
    
    # Define ChemBERTa model and move it to the specified device
    ChemBERTA = AutoModel.from_pretrained(pretrained_model_name_or_path=model_name).to(device)
    
    # Load custom tokenizer using the saved vocab.json
    custom_tokenizer = Tokenizer(
        WordLevel.from_file(
            'ChemBERTa/vocab.json',  # Path to your custom vocabulary file
            unk_token='[UNK]'
        )
    )

    # Set the pre-tokenizer to split SMILES characters (including handling Br, Cl, etc.)
    pre_tokenizer = Split(
        pattern=Regex(r"\[(.*?)\]|Br|Cl|."),
        behavior='isolated'
    )
    custom_tokenizer.pre_tokenizer = pre_tokenizer
    return ChemBERTA, custom_tokenizer
```

The key is to define a custom Regex.


We then use our own custum function to build the embedding:

```
def get_smiles_embedding(smiles, custom_tokenizer, ChemBERTA, device, max_length=512):
    # Tokenize the SMILES using your custom tokenizer
    custom_encoded = custom_tokenizer.encode(smiles)

    # Add [CLS] and [SEP] tokens
    CLS_token_id = 12  # Assuming 12 is the token ID for [CLS]
    SEP_token_id = 13  # Assuming 13 is the token ID for [SEP]
    PAD_token_id = 0   # Assuming 0 is the token ID for [PAD]

    # Create input_ids with [CLS] at the start and [SEP] at the end
    input_ids = [CLS_token_id] + custom_encoded.ids + [SEP_token_id]

    # Apply truncation if input_ids are longer than max_length
    if len(input_ids) > max_length:
        input_ids = input_ids[:max_length - 1] + [SEP_token_id]  # Ensure the sequence ends with [SEP]

    # Apply padding if input_ids are shorter than max_length
    if len(input_ids) < max_length:
        padding_length = max_length - len(input_ids)
        input_ids = input_ids + [PAD_token_id] * padding_length

    # Convert the input_ids to PyTorch tensor format
    input_ids_tensor = torch.tensor([input_ids]).to(device)

    # Prepare attention mask (1 for real tokens, 0 for padding)
    attention_mask = (input_ids_tensor != PAD_token_id).long()

    with torch.no_grad():
        # Get embeddings from ChemBERTA
        emb = ChemBERTA(
            input_ids=input_ids_tensor,
            attention_mask=attention_mask
        )["last_hidden_state"][:, 0, :].numpy()# Take [CLS] token embedding
    return emb
```


Here is the result for the SMILES "[O-][N+]1=CC=CC=C1"

```
[12, 36, 41, 20, 22, 16, 16, 22, 16, 16, 22, 16, 20, 13, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
Decoded Tokens: [CLS] [O-] [N+] 1 = C C = C C = C 1 [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]
tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0]])
```

Best,

Thomas

**Comment by DreRnc** [comment]
Hi @tspecht93, I agree with your regex method, as I ended up using MolFormer, which has a similar regex pattern, for my project. 

Just to answer regarding your hotfix though: I don't think this works as you would hope (as a "hotfix").

The model is trained on a certain vocabulary (and thus some numerical values associated with the different tokens), and I don't see why it should perform well (or perform at all) with a vocabulary of your choice, that you change after the model has been trained.

If you intend to retrain it instead, I get what you are doing!

**Comment by tspecht93** [comment]
Hi @DreRnc ,

We are aware of the fact, that Chemberta presumably was trained with the wrong tokenizer and you are right that the embeddings of certain atoms (e.g. "Cl") are not optimal. You are totally right that a retraining would be the "clean" way.

However, we still achieved a superior performance using the "Hotfix" instead of using the raw tokenizer. The reason (in my opinion, open to discussion :-)) is that the prediction head for the target property also has it's own parameters. We observed that these prediction head parameters are able to understand also these partially suboptimal embeddings.

Consider the case of CCCl. In the original tokenization, this would be [C C C]. Of course the prediction head will not be able to differentiate between "CCCl" and "CCC" with the wrong tokenizer. However, with the hotfix you will end up giving the prediction head unique embeddings for "CCC" and "CCCl" and thus you are able to correlate your target property quite well.

You can also use our Hotfix and fine-tune on your target property by unfreezing the ChemBERTa parameters. I think this should also work, but since it is much easier we did keep ChemBERTa as it is and only adjusted parameters in our parameters.

Thank you for pointing out MolFormer, will definitely take a look at it and how it performs in comparison to ChemBERTa.

Best,

Thomas
